core/templates/dev/head/pages/exploration_editor/editor_tab/SolutionVerificationServiceSpec.js

```diff
@@ -17,183 +17,204 @@
+    var interactionCtrl = $controller('StateInteraction', {
+      $scope: scope,
+      editorContextService: ecs,
+      changeListService: cls,
+      explorationStatesService: ess,
+      editabilityService: {
+        isEditable: function() {
+          return true;
+        }
+      },
+      stateInteractionIdService: siis,
+      stateCustomizationArgsService: scas,
+      interactionDetailsCache: idc,
+      INTERACTION_SPECS: IS
+    });
+  }));
 
-      var interactionCtrl = $controller('StateInteraction', {
-        $scope: scope,
-        editorContextService: ecs,
-        changeListService: cls,
-        explorationStatesService: ess,
-        editabilityService: {
-          isEditable: function() {
-            return true;
-          }
+  it('should verify a correct solution',
+    function(done) {
+      ecs.setActiveStateName('First State');
+      scope.initStateEditor();
+      var state = ess.getState('First State');
+      siis.init(
+        'First State', state.interaction.id, state.interaction, 'widget_id');
+      scas.init(
+        'First State', state.interaction.customizationArgs,
+        state.interaction, 'widget_customization_args');
+
+      siis.displayed = 'TextInput';
+      scope.onCustomizationModalSavePostHook();
+
+      ess.saveSolution('First State', sof.createNew(false, 'abc', 'nothing'));
+
+      svs.verifySolution(0, state,
+        ess.getState('First State').interaction.solution.correctAnswer,
+        function() {
+          ess.updateSolutionValidity('First State', true);
         },
-        stateInteractionIdService: siis,
-        stateCustomizationArgsService: scas,
-        interactionDetailsCache: idc,
-        INTERACTION_SPECS: IS
-      });
-    }));
-
-    it('should verify the solution correctly',
-      function() {
-        ecs.setActiveStateName('First State');
-        scope.initStateEditor();
-
-        var state = ess.getState('First State');
-        siis.init(
-          'First State', state.interaction.id, state.interaction, 'widget_id');
-        scas.init(
-          'First State', state.interaction.customizationArgs,
-          state.interaction, 'widget_customization_args');
-
-        siis.displayed = 'TextInput';
-        scope.onCustomizationModalSavePostHook();
-
-        ess.saveSolution('First State', sof.createNew(false, 'abc', 'nothing'));
-
-        svs.verifySolution(0, state,
-          ess.getState('First State').interaction.solution.correctAnswer,
-          function() {
-            ess.updateSolutionValidity('First State', true);
-            expect(ess.isSolutionValid('First State')).toBe(true);
-          },
-          function() {
-            ess.updateSolutionValidity('First State', false);
-          });
+        function() {
+          ess.updateSolutionValidity('First State', false);
+        }
+        ).then(function() {
+          expect(ess.isSolutionValid('First State')).toBe(true);
+          done();
+        });
+      scope.$apply();
```

Comment:
```
it('should simulate promise', inject(function($q, $rootScope) {
  var deferred = $q.defer();
  var promise = deferred.promise;
  var resolvedValue;

  promise.then(function(value) { resolvedValue = value; });
  expect(resolvedValue).toBeUndefined();

  // Simulate resolving of promise
  deferred.resolve(123);
  // Note that the 'then' function does not get called synchronously.
  // This is because we want the promise API to always be async, whether or not
  // it got called synchronously or asynchronously.
  expect(resolvedValue).toBeUndefined();

  // Propagate promise resolution to 'then' functions using $apply().
  $rootScope.$apply();
  expect(resolvedValue).toEqual(123);
}));
```
Source: https://docs.angularjs.org/api/ng/service/$q